### PR TITLE
[PREVIEW] ROC-3257: Rework states paid defence flow

### PR DESCRIPTION
**Pull request by andrewwa-kainos**
### JIRA link
https://tools.hmcts.net/jira/browse/ROC-3257

### Change description
Defendant can defend a claim by informing the claimant that they have paid, amount, when, and how.

### Work checklist

- [x] Unit tests added where applicable
- [x] Route tests added for new pages
- [x] New pages included in a11y tests
- [x] Task list and task completeness checks updated
- [x] Check and send page updated
- [x] Routes, page content and page flows of new features are toggled 
- [x] UI changes look good on mobile
- [ ] Required Google Analytics events are being sent 

### Developer self-QA run statement
- [x] I have clicked through the running application to see if all changes I made actually work.

### Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

File: src/main/features/response/draft/responseDraft.ts [hunk 1]
```diff
@@ -187,13 +187,19 @@ export class ResponseDraft extends DraftDocument {
   }
 
   public isResponseRejected (): boolean {
-    if (!this.isResponsePopulated()) {
+    if (!this.response || !this.response.type) {
```
Comment (timja) [inline src/main/features/response/draft/responseDraft.ts]: isn't this exactly what `isResponsePopulated` does?

File: src/main/features/response/routes/how-much-have-you-paid.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import * as express from 'express'
```
Comment (andrewwa-kainos) [inline src/main/features/response/routes/how-much-have-you-paid.ts]: duplicate, more or less, of that in partial-admission

File: src/main/features/response/routes/how-much-have-you-paid.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    totalAmount: res.locals.claim.totalAmountTillToday,
+    pastDate
+  })
+}
+
+/* tslint:disable:no-default-export */
+export default express.Router()
+  .get(
+    page.uri,
+    ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction) => {
```
Comment (andrewwa-kainos) [inline src/main/features/response/routes/how-much-have-you-paid.ts]: other guards needed?

File: src/main/features/response/routes/why-do-you-disagree.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import * as express from 'express'
```
Comment (andrewwa-kainos) [inline src/main/features/response/routes/why-do-you-disagree.ts]: duplicate, more or less, of that in partial-admission

File: src/main/features/response/routes/you-have-paid-less.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import * as express from 'express'
+import { Paths } from 'response/paths'
+
+async function renderView (res: express.Response, next: express.NextFunction) {
+  try {
```
Comment (timja) [inline src/main/features/response/routes/you-have-paid-less.ts]: why do you have a try catch here?
i can't see anything that could fail

File: src/main/features/response/views/check-and-send.njk [hunk 6]
```diff
@@ -15,38 +15,34 @@
     {{ csrfProtection(csrf) }}
```
Comment (andrewwa-kainos) [inline src/main/features/response/views/check-and-send.njk]: lots of auto-formatting changes in this file

File: src/main/features/response/views/how-much-have-you-paid.njk [hunk 7]
```diff
@@ -0,0 +1,31 @@
+{% extends "layout.njk" %}
```
Comment (andrewwa-kainos) [inline src/main/features/response/views/how-much-have-you-paid.njk]: duplicate, more or less, of that in partial-admission

File: src/main/features/response/views/why-do-you-disagree.njk [hunk 8]
```diff
@@ -0,0 +1,22 @@
+{% extends "layout.njk" %}
```
Comment (andrewwa-kainos) [inline src/main/features/response/views/why-do-you-disagree.njk]: duplicate, more or less, of that in partial-admission

File: src/main/features/response/views/check-and-send.njk [hunk 9]
```diff
@@ -15,38 +15,34 @@
...
 
     {{ tableStart('yourDetails', 'Your details') }}
-    {{
```
Comment (timja) [inline src/main/features/response/views/check-and-send.njk]: changes like this are big conflicts risk, can you undo and just apply your changes on top?

File: src/test/features/response/routes/how-much-have-you-paid.ts [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+          })
+          await request(app)
+            .get(pagePath)
+            .set('Cookie', `${cookieName}=ABC`)
+            .expect(res => expect(res).to.be.successful.withText(header))
+        })
+      })
+    })
+
+    describe('on POST', () => {
```
Comment (andrewwa-kainos) [inline src/test/features/response/routes/how-much-have-you-paid.ts]: needs pulled up out of 'on GET'

File: src/test/features/response/routes/why-do-you-disagree.ts [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+          })
+          await request(app)
+            .get(pagePath)
+            .set('Cookie', `${cookieName}=ABC`)
+            .expect(res => expect(res).to.be.successful.withText(header))
+        })
+      })
+    })
+
+    describe('on POST', () => {
```
Comment (andrewwa-kainos) [inline src/test/features/response/routes/why-do-you-disagree.ts]: needs pulled out of 'on GET'

File: src/main/features/response/draft/responseDraft.ts (hunk 1 above)
Comment (andrewwa-kainos) [reply src/main/features/response/draft/responseDraft.ts]: I thought so too, but changing it fixed a test... Will revisit

File: src/main/features/response/draft/responseDraft.ts [hunk 12]
```diff
@@ -194,6 +194,12 @@ export class ResponseDraft extends DraftDocument {
     return this.response.type === ResponseType.DEFENCE
   }
 
+  public isResponseRejectedFullyBecausePaidInFull (): boolean {
```
Comment (robertparkinson) [inline src/main/features/response/draft/responseDraft.ts]: this seems very similar to isResponseRejectedFullyWithAmountClaimedPaid

**Comment by andrewwa-kainos** [comment]
Missing feature toggle

File: src/main/features/response/routes/full-rejection/why-do-you-disagree.ts [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+      const form: Form<WhyDoYouDisagree> = req.body
+
+      if (form.hasErrors()) {
+        renderView(form, res)
+      } else {
+        const draft: Draft<ResponseDraft> = res.locals.responseDraft
+        const user: User = res.locals.user
+
+        draft.document.rejectAllOfClaim.whyDoYouDisagree = form.model
+        draft.document.partialAdmission = draft.document.fullAdmission = undefined
```
Comment (geek-soft) [inline src/main/features/response/routes/full-rejection/why-do-you-disagree.ts]: That should not be here. Not right place to do so. Should be done on page with 3 response options

File: src/main/features/response/routes/full-rejection/you-have-paid-less.ts [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+  .get(FullRejectionPaths.youHavePaidLess.uri,
+    FullRejectionGuard.requestHandler(),
+    OptInFeatureToggleGuard.featureEnabledGuard('admissions'),
+    async (req: express.Request, res: express.Response, next: express.NextFunction) => {
+      await Promise.resolve(res.render(FullRejectionPaths.youHavePaidLess.associatedView, { claim: res.locals.claim }))
+    })
+  .post(FullRejectionPaths.youHavePaidLess.uri,
+    FullRejectionGuard.requestHandler(),
+    OptInFeatureToggleGuard.featureEnabledGuard('admissions'),
+    function (req, res) {
```
Comment (geek-soft) [inline src/main/features/response/routes/full-rejection/you-have-paid-less.ts]: Please add types to req and res.

File: src/main/features/response/routes/full-rejection/you-have-paid-less.ts [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+      res.redirect(Paths.taskListPage.evaluateUri({ externalId: res.locals.claim.externalId }))
```
Comment (geek-soft) [inline src/main/features/response/routes/full-rejection/you-have-paid-less.ts]: Please extract claim to local variable to give it type.

File: src/main/features/response/routes/reject-all-of-claim.ts [hunk 16]
```diff
@@ -67,7 +68,11 @@ export default express.Router()
             res.redirect(Paths.sendYourResponseByEmailPage.evaluateUri({ externalId: externalId }))
             break
           case RejectAllOfClaimOption.ALREADY_PAID:
-            res.redirect(Paths.defendantHowMuchPaidClaimant.evaluateUri({ externalId: externalId }))
+            if (FeatureToggles.isEnabled('admissions')) {
```
Comment (geek-soft) [inline src/main/features/response/routes/reject-all-of-claim.ts]: That only checks config, please use one that checks opted in features on claim. See `ClaimFeatureToggles.areAdmissionsEnabled`

File: src/main/features/response/tasks/howMuchHaveYouPaidTask.ts [hunk 17]
```diff
@@ -6,7 +6,13 @@ const validator = new Validator()
 
 export class HowMuchHaveYouPaidTask {
   static isCompleted (responseDraft: ResponseDraft): boolean {
```
Comment (geek-soft) [inline src/main/features/response/tasks/howMuchHaveYouPaidTask.ts]: That task should only take `howMuchHaveYouPaid`. Then task list builder, depending on what task is being used, passes right instance for validation.

In the future we will remove task classes, similar to HowMuchHaveYouPaidTask, completely as they only call `!!model && validator.validateSync(model).length === 0` in most of the cases

File: src/main/features/response/tasks/whyDoYouDisagreeTask.ts [hunk 18]
```diff
@@ -6,7 +6,12 @@ const validator = new Validator()
 
 export class WhyDoYouDisagreeTask {
   static isCompleted (responseDraft: ResponseDraft): boolean {
-    return WhyDoYouDisagreeTask.isWhyDoYouDisagreeValid(responseDraft.partialAdmission.whyDoYouDisagree)
+    if (responseDraft.isResponsePartiallyAdmitted()) {
```
Comment (geek-soft) [inline src/main/features/response/tasks/whyDoYouDisagreeTask.ts]: The same here.

File: src/main/features/response/views/check-and-send.njk [hunk 19]
```diff
@@ -63,6 +63,38 @@
       {% endif %}
 
       {{ row(label = 'Why do you reject all of this claim?', value = rejectAllOption, changeLink = ResponsePaths.defenceRejectAllOfClaimPage.evaluateUri({ externalId: claim.externalId }), bold = true) }}
+
+      {% if draft.rejectAllOfClaim.option === RejectAllOfClaimOption.ALREADY_PAID %}
+        {{
+          row(
+            label = 'How much have you paid?',
+            value = t(draft.rejectAllOfClaim.howMuchHaveYouPaid.amount | numeral),
```
Comment (geek-soft) [inline src/main/features/response/views/check-and-send.njk]: Amounts do not need to be passed through `t()`. There are same in English and Welsh

File: src/main/features/response/views/check-and-send.njk [hunk 20]
```diff
@@ -63,6 +63,38 @@
...
+            value = t(draft.rejectAllOfClaim.howMuchHaveYouPaid.date.toMoment() | date),
+            bottomBorder = false,
+            bold = true
+          )
+        }}
+
+        {{
+          row(
+            label = 'How did you pay?',
+            value = t(draft.rejectAllOfClaim.howMuchHaveYouPaid.text),
```
Comment (geek-soft) [inline src/main/features/response/views/check-and-send.njk]: User entered text cannot be passed through `t()` as it varies from case to case.

File: src/main/features/response/views/check-and-send.njk [hunk 21]
```diff
@@ -63,6 +63,38 @@
...
+            bold = true
+          )
+        }}
+
+        {% if draft.rejectAllOfClaim.howMuchHaveYouPaid.amount < claim.totalAmountTillToday %}
+          {{ row(label = 'Why do you disagree with the amount claimed?', value = t(draft.rejectAllOfClaim.whyDoYouDisagree.text), changeLink = FullRejectionPaths.whyDoYouDisagreePage.evaluateUri({ externalId: claim.externalId }), bold = true) }}
```
Comment (geek-soft) [inline src/main/features/response/views/check-and-send.njk]: The same here, do no pass `draft.rejectAllOfClaim.whyDoYouDisagree.text` though `t()`

File: src/main/features/response/views/full-rejection/how-much-have-you-paid.njk [hunk 22]
```diff
@@ -0,0 +1,31 @@
+{% extends "layout.njk" %}
```
Comment (geek-soft) [inline src/main/features/response/views/full-rejection/how-much-have-you-paid.njk]: Nasty duplicate :( Can you please create story to get rid of it after Wed?

File: src/main/features/response/views/full-rejection/why-do-you-disagree.njk [hunk 23]
```diff
@@ -0,0 +1,22 @@
+{% extends "layout.njk" %}
+{% from "form.njk" import csrfProtection, textArea, saveAndContinueButton, errorSummary %}
+
+{% set heading = 'Why do you disagree with the claim amount?' %}
```
Comment (geek-soft) [inline src/main/features/response/views/full-rejection/why-do-you-disagree.njk]: The same here, please story to remove duplicates with list of these we are introducing now. Mark that story with tech debt label.

File: src/main/features/response/views/check-and-send.njk (hunk 19 above)
Comment (andrewwa-kainos) [reply src/main/features/response/views/check-and-send.njk]: Ahh, that's what `t()` does!

File: src/main/features/response/views/full-rejection/why-do-you-disagree.njk (hunk 23 above)
Comment (andrewwa-kainos) [reply src/main/features/response/views/full-rejection/why-do-you-disagree.njk]: ROC-4282

File: src/main/features/response/views/full-rejection/how-much-have-you-paid.njk (hunk 22 above)
Comment (andrewwa-kainos) [reply src/main/features/response/views/full-rejection/how-much-have-you-paid.njk]: ROC-4282

File: src/main/features/response/routes/full-rejection/how-much-have-you-paid.ts [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+
+        if (!paidLessThanClaimed) {
+          delete draft.document.rejectAllOfClaim.whyDoYouDisagree
+          delete draft.document.timeline
+          delete draft.document.evidence
+          delete draft.document.freeMediation
+        }
+
+        await new DraftService().save(draft, user.bearerToken)
+        /*
```
Comment (timja) [inline src/main/features/response/routes/full-rejection/how-much-have-you-paid.ts]: can this be removed? it confuses me more than reading the code

File: src/main/features/response/views/check-and-send.njk [hunk 25]
```diff
@@ -200,6 +194,63 @@
       {% endif %}
     {% endif %}
 
+    {% if draft.rejectAllOfClaim.option === RejectAllOfClaimOption.ALREADY_PAID %}
+      {{ tableStart('responseDetails', 'Response details') }}
+      {{
```
Comment (timja) [inline src/main/features/response/views/check-and-send.njk]: indenting is off here

File: src/main/features/response/views/check-and-send.njk [hunk 26]
```diff
@@ -235,7 +286,8 @@
       {% include "./statement-of-means/check-and-send.njk" %}
     {% endif %}
 
-    {% if (admissionsApplicable and draft.isResponsePartiallyAdmitted()) or draft.isResponseRejectedFullyWithDispute() %}
+    {% if (admissionsApplicable and (draft.isResponsePartiallyAdmitted() or (draft.isResponseRejectedFullyBecausePaidWhatOwed() and draft.rejectAllOfClaim.howMuchHaveYouPaid.amount < claim.totalAmountTillToday)))
```
Comment (timja) [inline src/main/features/response/views/check-and-send.njk]: unneeded brackets after if

File: src/main/features/response/views/check-and-send.njk (hunk 25 above)
Comment (geek-soft) [reply src/main/features/response/views/check-and-send.njk]: Done.

File: src/main/features/response/routes/full-rejection/how-much-have-you-paid.ts (hunk 24 above)
Comment (geek-soft) [reply src/main/features/response/routes/full-rejection/how-much-have-you-paid.ts]: Done

**Pull request merged**